src/Nethermind/Nethermind.Merge.Plugin/InvalidChainTracker/InvalidChainTracker.cs

```diff
@@ -0,0 +1,188 @@
+// 
+
+using System;
+using System.Collections.Generic;
+using Nethermind.Blockchain;
+using Nethermind.Blockchain.Find;
+using Nethermind.Consensus;
+using Nethermind.Consensus.Processing;
+using Nethermind.Core;
+using Nethermind.Core.Caching;
+using Nethermind.Core.Crypto;
+using Nethermind.Logging;
+using Nethermind.Merge.Plugin.Handlers;
+
+namespace Nethermind.Merge.Plugin.InvalidChainTracker;
+
+/// <summary>
+/// Something that tracks if a given hash is on a known invalid chain, as one if it's ancestor have been reported to
+/// be invalid.
+/// 
+/// </summary>
+public class InvalidChainTracker: IInvalidChainTracker
+{
+    private readonly IPoSSwitcher _poSSwitcher;
+    private readonly IBlockFinder _blockFinder;
+    private readonly IBlockCacheService _blockCacheService;
+    private readonly ILogger _logger;
+    private readonly object _opLock = new();
+    private readonly LruCache<Keccak, Node> _tree;
+    
+    // CompositeDisposable only available on System.Reactive. So this will do for now. 
+    private readonly List<Action> _disposables = new();
+
+    public InvalidChainTracker(
+        IPoSSwitcher poSSwitcher, 
+        IBlockFinder blockFinder,
+        IBlockCacheService blockCacheService,
+        ILogManager logManager)
+    {
+        _poSSwitcher = poSSwitcher;
+        _blockFinder = blockFinder;
+        _tree = new LruCache<Keccak, Node>(1024, nameof(InvalidChainTracker));
+        _logger = logManager.GetClassLogger<InvalidChainTracker>();
+        _blockCacheService = blockCacheService;
+    }
+
+    public void SetupBlockchainProcessorInterceptor(IBlockchainProcessor blockchainProcessor)
+    {
+        blockchainProcessor.OnInvalidBlock += BlockchainProcessorOnOnInvalidBlock;
+        _disposables.Add(() =>
+        {
+            blockchainProcessor.OnInvalidBlock -= BlockchainProcessorOnOnInvalidBlock;
+        });
+    }
+
+    private void BlockchainProcessorOnOnInvalidBlock(object? sender, InvalidBlockException e)
+    {
+        OnInvalidBlock(e.InvalidBlockHash, null);
+    }
+
+    public void SetChildParent(Keccak child, Keccak parent)
+    {
+        lock (_opLock)
+        {
+            Node parentNode = GetNode(parent);
+
+            parentNode.Children.Add(child);
+            if (parentNode.LastValidHash != null)
+            {
+                PropagateLastValidHash(parentNode);
+            }
+        }
+    }
+
+    private Node GetNode(Keccak hash)
+    {
+        if (!_tree.TryGet(hash, out Node node))
+        {
+            node = new Node();
+            _tree.Set(hash, node);
+        }
+
+        return node;
+    }
+
+    private void PropagateLastValidHash(Node node)
+    {
+        foreach (Keccak nodeChild in node.Children)
+        {
+            Node childNode = GetNode(nodeChild);
+            if (childNode.LastValidHash != node.LastValidHash)
+            {
+                childNode.LastValidHash = node.LastValidHash;
+                PropagateLastValidHash(childNode);
```

Comment:
Potentially 1024 (or more if we increase cache size) deep recursion!

Reply:
Yes, I know. It could have 1024 descendant. Line 1024 should prevent child which have been marked from being traversed again.

Reply:
I could change it to a BFS, which remove the recursion if you want.

Reply:
Yes, but I would also advise not to track children but track parent, see next comment.